### PR TITLE
commons #34 Iterator.copyToArray alternative that returns a number of…

### LINKED ISSUE #34: Iterator.copyToArray alternative that returns a number of copies items
The standard Scala method `Iterator.copyToArray()` returns `Unit`.
It often is useful to know how many items has been actually copied over / read from an iterator, especially when writing a portion into a bigger array.

### PR DESCRIPTION
**Pull request by wajda**
… copies items

fixes #34

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -11,6 +11,18 @@ Selection of useful reusable components
...
+# Collection utils
+
+```scala
+import CollectionImplicits._
+
+val iter: Iterator[_] = ???
+val arr: Array[_] = ???
+
+// copy 42 items from the `iter` to the `arr` with array offset 7
+iter.fetch(arr, 7, 42) // returns a number of actually copied items
```
Comment (cerveada) [inline README.md:23]: To me fetch just doesn't sound right. I would expect the iterator to fetch me something for example return the array. 

Why not use `copyWithCount` ?

Comment (wajda) [reply README.md:23]: The original method is called `copyToArray`.  So what about `fetchToArray` then?

I'm not keen on using `WithCount` in this context as it sounds like you "copy items with count", while in fact you only copy items along, and then return they count. So personally I'd avoid mixing what a method does (a side-effect) with what it returns as a function in one name.

Comment (cerveada) [reply README.md:23]: `fetchToArray` sounds good.

**Review by cerveada — APPROVED** [review]

**Review by yruslan — APPROVED** [review]

File: src/main/scala/za/co/absa/commons/lang/CollectionImplicits.scala [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.commons.lang
+
+object CollectionImplicits {
+
+  implicit class IteratorOps[A](iter: Iterator[A]) {
```
Comment (cerveada) [inline src/main/scala/za/co/absa/commons/lang/CollectionImplicits.scala:21]: Could this be made into value class? That would avoid creation of unnecessary instances durring runtime.

https://docs.scala-lang.org/overviews/core/value-classes.html#extension-methods

Comment (wajda) [reply src/main/scala/za/co/absa/commons/lang/CollectionImplicits.scala]: done

File: src/main/scala/za/co/absa/commons/lang/CollectionImplicits.scala [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+      * that returns the number of actually read elements.
+      *
+      * @param xs    the array to fill.
+      * @param start the starting index.
+      * @param len   the maximal number of elements to copy.
+      * @tparam B the type of the elements of the array.
+      * @note Reuse: $consumesIterator
+      */
+    def fetch[B >: A](xs: Array[B], start: Int, len: Int): Int = {
+      require(start >= 0 && (start < xs.length || xs.length == 0), s"start $start out of range ${xs.length}")
```
[the remark this answers is not in the document]
Comment (wajda) [reply src/main/scala/za/co/absa/commons/lang/CollectionImplicits.scala:33]: to comply with the contract of the original Scala method - `Iterator.copyToArray()`

**Review by cerveada — DISMISSED** [review]

File: src/main/scala/za/co/absa/commons/lang/CollectionImplicits.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (benedeki) [reply src/main/scala/za/co/absa/commons/lang/CollectionImplicits.scala:33]: Fair enough

**Review by yruslan — APPROVED** [review]

**Pull request merged**